### PR TITLE
Add settings to change local port

### PR DESCRIPTION
**Pull request by Atlante45**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: f791b8de1c64576074b4d90139867a878f846d0f available at: https://gist.github.com/d36cf04dad44bdb151a7d2e9cf9b5778<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8513-f791b8d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8513-f791b8d.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f6d71dfb4eb0351fae4558bad802b2133758903f available at: https://gist.github.com/d36cf04dad44bdb151a7d2e9cf9b5778<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8513-f6d71df.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8513-f6d71df.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 35b2e69e6d9bbfb0dfe8f2c5cbadeec0fe033341 available at: https://gist.github.com/d36cf04dad44bdb151a7d2e9cf9b5778<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8513-35b2e69.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8513-35b2e69.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3abf342edaf13b10e9e16b9eda85c7d4bf93cfb4 available at: https://gist.github.com/d36cf04dad44bdb151a7d2e9cf9b5778<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8513-3abf342.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8513-3abf342.dmg

**Comment by OmegaHeron** [comment]
@Atlante45 What's testing plan on this?

**Comment by Atlante45** [comment]
Test plan:

While running the PR's Sandbox:

- Navigate to `Developer > Network > Networking...`
- Set the port to something else and save.
- Wait for the domain to fully reload.
- After restart, the interface should still be using the same port.

- Start interface while providing a new port via command line argument. (ie. `--listenPort XXXX`)
       This should override the port used, but only for this session.

- Check that using the port 0 will result in a random port.


- Make sure you can still connect correctly to other domain, and that restarting a domain while connected to it does not pose any issues.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 766819511768994b03de46203259213a70d7f978 available at: https://gist.github.com/d36cf04dad44bdb151a7d2e9cf9b5778<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8513-7668195.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8513-7668195.dmg

**Comment by OmegaHeron** [comment]
Working on this now - just need to get stack compiled for Linux first.

**Comment by OmegaHeron** [comment]
Testing Interface on Windows 10 - Stack on Ubuntu Linux 16.04.1

QA Fail.  If I change port and hit save - nothing happens.  It remains at port it was originally set to.

Log file shows two things that may be of help;

When opening network setting

`[08/24 16:03:50] [WARNING] file:///C:/Program Files/High Fidelity - PR8513/resources/qml/dialogs/preferences/SpinBoxPreference.qml:54:23: Unable to assign [undefined] to int`

After changing and applying

``[08/24 16:04:12] [WARNING] QMetaObject::invokeMethod: No such method NodeList::setSocketLocalPort(quint16)`

**Comment by hifi-gustavo** [comment]
Build console output for commit: 12eb65c0c4b06b2debcb442aeb1f1f3e1d06d514 available at: https://gist.github.com/d36cf04dad44bdb151a7d2e9cf9b5778<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8513-12eb65c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8513-12eb65c.dmg

**Comment by Atlante45** [comment]
@OmegaHeron Sorry for the delay, I was OOO.
I moved the setter to the base class and forgot to make it invokable.
It should be good now.

**Comment by OmegaHeron** [comment]
Testing Windows 10


Navigate to Developer > Network > Networking...
Set the port to something else and save.
Wait for the domain to fully reload.
After restart, the interface should still be using the same port.

**Yes to all above**

Start interface while providing a new port via command line argument. (ie. --listenPort XXXX)
This should override the port used, but only for this session.

**Yes - I had set port to 59222 from inside Interface.  When I passed --listenPort 59252 via cmd it used 59256.  When I launched again without port flag it used 59222**

Check that using the port 0 will result in a random port.

**Yes setting port = 0 within Interface GUI or via command line results in randomized port**
**NO setting port = 0 via cmd --listenPort 0 does not use random - it used the manually saved port of 59222**

Make sure you can still connect correctly to other domain

**Yes - I tried 4 other domains and all were good.**

and that restarting a domain while connected to it does not pose any issues.

**Yes - Restarting server connected to results in Interface reconnecting once stack returns.**

I'm not sure if this is QA Pass or Fail.  If setting listenPort to 0 from cmd line isn't intended to create random port QA=Pass else QA=Fail.

**Comment by Atlante45** [comment]
@OmegaHeron Great find, this is due to the way we signal "No port was passed as an argument", by using the value 0.
Gonna fix that.

**Comment by hifi-gustavo** [comment]
Build console output for commit: eb007b24cdb7b6db2f42fd92ecb4657b1bd8f48c available at: https://gist.github.com/d36cf04dad44bdb151a7d2e9cf9b5778<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8513-eb007b2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8513-eb007b2.dmg

**Comment by OmegaHeron** [comment]
QA Pass

Windows 10

Setting port on cmd line to 0 gives random, setting to > 0 gives requested (Unless on Linux and using low port values and !root ;) )

All permutations of test list work as expected.

File: libraries/shared/src/Preferences.h [hunk 1]
```diff
@@ -189,6 +189,38 @@ class FloatPreference : public TypedPreference<float> {
...
+class IntPreference : public TypedPreference<int> {
+    Q_OBJECT
+    Q_PROPERTY(float value READ getValue WRITE setValue NOTIFY valueChanged)
+    Q_PROPERTY(float min READ getMin CONSTANT)
+    Q_PROPERTY(float max READ getMax CONSTANT)
+    Q_PROPERTY(float step READ getStep CONSTANT)
+
+public:
+    IntPreference(const QString& category, const QString& name, Getter getter, Setter setter)
+    : TypedPreference(category, name, getter, setter) { }
```
Comment (birarda) [inline libraries/shared/src/Preferences.h]: tab this over

**Comment by hifi-gustavo** [comment]
Build console output for commit: 006780fe8e89103c15400bbe33153cbadfe47e5d available at: https://gist.github.com/d36cf04dad44bdb151a7d2e9cf9b5778<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8513-006780f.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8513-006780f.dmg

**Pull request merged**